server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/MockDiskUsagesIT.java

```diff
@@ -16,215 +16,416 @@
 
-        // Wait until index "test" is blocked
-        assertBusy(() -> assertBlocked(client().prepareIndex().setIndex("test").setType("doc").setId("1").setSource("foo", "bar"),
+        assertBusy(() -> assertBlocked(
+            client().prepareIndex().setIndex("test").setType("doc").setId("1").setSource("foo", "bar"),
             IndexMetaData.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK));
 
         assertFalse(client().admin().cluster().prepareHealth("test").setWaitForEvents(Priority.LANGUID).get().isTimedOut());
 
         // Cannot add further documents
         assertBlocked(client().prepareIndex().setIndex("test").setType("doc").setId("2").setSource("foo", "bar"),
             IndexMetaData.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK);
-        assertSearchHits(client().prepareSearch().get(), "1");
+        assertSearchHits(client().prepareSearch("test").get(), "1");
+
+        logger.info("--> index is confirmed read-only, releasing disk space");
 
-        // Update the disk usages so all nodes are back under the high and flood watermarks
-        cis.setN1Usage(realNodeNames.get(0), new DiskUsage(nodes.get(0), "n1", "_na_", 100, 11));
-        cis.setN2Usage(realNodeNames.get(1), new DiskUsage(nodes.get(1), "n2", "_na_", 100, 11));
-        cis.setN3Usage(realNodeNames.get(2), new DiskUsage(nodes.get(2), "n3", "_na_", 100, 11));
+        // Move all nodes below the high watermark so that the index is unblocked
+        clusterInfoService.diskUsageFunction = (discoveryNode, fsInfoPath) -> setDiskUsage(fsInfoPath, 100, between(10, 100));
 
         // Attempt to create a new document until DiskUsageMonitor unblocks the index
         assertBusy(() -> {
             try {
-                client().prepareIndex("test", "doc", "3").setSource("{\"foo\": \"bar\"}", XContentType.JSON)
+                client().prepareIndex("test", "doc", "3").setSource("foo", "bar")
                     .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
             } catch (ClusterBlockException e) {
                 throw new AssertionError("retrying", e);
             }
         });
-        assertSearchHits(client().prepareSearch().get(), "1", "3");
+        assertSearchHits(client().prepareSearch("test").get(), "1", "3");
+    }
+
+    public void testOnlyMovesEnoughShardsToDropBelowHighWatermark() throws Exception {
+        for (int i = 0; i < 3; i++) {
+            // ensure that each node has a single data path
+            internalCluster().startNode(Settings.builder().put(Environment.PATH_DATA_SETTING.getKey(), createTempDir()));
+        }
+
+        final MockInternalClusterInfoService clusterInfoService = (MockInternalClusterInfoService)
+            internalCluster().getInstance(ClusterInfoService.class, internalCluster().getMasterName());
+
+        final AtomicReference<ClusterState> masterAppliedClusterState = new AtomicReference<>();
+        internalCluster().getInstance(ClusterService.class, internalCluster().getMasterName())
```

Comment:
Why not use `internalCluster().getCurrentMasterNodeInstance()`?

Reply:
TIL :) b8d9d1ecebe25325cc20468895fd4b6d450f2d29